File: src/com/github/gtache/client/ProcessStreamConnectionProvider.scala
/** *****************************************************************************
  * Copyright (c) 2016 Rogue Wave Software Inc. and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
  * http://www.eclipse.org/legal/epl-v10.html
  *
  * Contributors:
  * Michał Niewrzał (Rogue Wave Software Inc.) - initial implementation
  * ******************************************************************************/
package com.github.gtache.client

import java.io.{File, IOException, InputStream, OutputStream}
import java.util.Objects

import com.intellij.openapi.diagnostic.Logger
import org.jetbrains.annotations.Nullable

class ProcessStreamConnectionProvider(private var commands: Seq[String], private var workingDir: String) extends StreamConnectionProvider {
  @Nullable private var process: Process = _
  private val LOG: Logger = Logger.getInstance(classOf[ProcessOverSocketStreamConnectionProvider])

  @throws[IOException]
  override def start(): Unit = {
    if (this.workingDir == null || this.commands == null || this.commands.isEmpty || this.commands.contains(null)) throw new IOException("Unable to start language server: " + this.toString) //$NON-NLS-1$
    val builder = createProcessBuilder
    LOG.info("Starting server process with commands " + commands + " and workingDir " + workingDir)
    this.process = builder.start

    if (!process.isAlive) throw new IOException("Unable to start language server: " + this.toString) else LOG.info("Server process started " + process)
  }

  protected def createProcessBuilder: ProcessBuilder = {
    import scala.collection.JavaConverters._
    val builder = new ProcessBuilder(getCommands.asJava)
    builder.directory(new File(getWorkingDirectory))
    builder.redirectError(ProcessBuilder.Redirect.INHERIT)
    builder
  }

  @Nullable override def getInputStream: InputStream = {
    if (process == null) null
    else process.getInputStream
  }

  @Nullable override def getOutputStream: OutputStream = {
    if (process == null) null
    else process.getOutputStream
  }

  override def stop(): Unit = {
    if (process != null) process.destroy()
  }

  override def equals(obj: Any): Boolean = {
    obj match {
      case other: ProcessStreamConnectionProvider =>
        getCommands.size == other.getCommands.size && this.getCommands.toSet == other.getCommands.toSet && this.getWorkingDirectory == other.getWorkingDirectory
      case _ => false
    }

  }

  override def hashCode: Int = {
    Objects.hashCode(this.getCommands) ^ Objects.hashCode(this.getWorkingDirectory)
  }

  protected def getCommands: Seq[String] = commands

  def setCommands(commands: Seq[String]): Unit = {
    this.commands = commands
  }

  protected def getWorkingDirectory: String = workingDir

  def setWorkingDirectory(workingDir: String): Unit = {
    this.workingDir = workingDir
  }
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/client/connection/ProcessStreamConnectionProvider.kt
/* Adapted from lsp4e*/
package com.github.gtache.lsp.client.connection

import com.intellij.openapi.diagnostic.Logger
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.*

/**
 * A class representing a connection to a process
 *
 * @param commands   The commands to start the process
 * @param workingDir The working directory of the process
 */
open class ProcessStreamConnectionProvider(private var commands: List<String>, private var workingDir: String) :
    StreamConnectionProvider {

    companion object {
        private val logger: Logger = Logger.getInstance(ProcessStreamConnectionProvider::class.java)
    }

    private var process: Process? = null

    override val inputStream: InputStream?
        get() = process?.inputStream
    override val errorStream: InputStream?
        get() = process?.errorStream
    override val outputStream: OutputStream?
        get() = process?.outputStream

    override fun start(): Unit {
        if (this.commands.isEmpty()) {
            throw IOException("Unable to start language server: $this")
        }
        val builder = createProcessBuilder()
        logger.info("Starting server process , commands ${commands.joinToString(" ")} and workingDir $workingDir")
        this.process = builder.start()
        process?.let {
            if (!it.isAlive) throw IOException("Unable to start language server: $this") else logger.info("Server process started $process")
        }
    }

    private fun createProcessBuilder(): ProcessBuilder {
        val builder = ProcessBuilder(commands.map { s -> s.replace("\'", "") })
        builder.directory(File(workingDir))
        return builder
    }

    override fun stop(): Unit {
        process?.destroy()
    }

    override fun equals(other: Any?): Boolean {
        return other is ProcessStreamConnectionProvider && commands.size == other.commands.size &&
                this.commands == other.commands && this.workingDir == other.workingDir

    }

    override fun hashCode(): Int {
        return Objects.hash(this.commands, this.workingDir)
    }
}